build: enable lintlua for test/ dir

justinmk opened the pull request:
Problem:
Not all Lua code is checked by stylua. Automating code-style is an important mechanism for reducing time spent on accidental (non-essential) complexity.

Solution:
- Enable stylua for entire `test/` directory.
- Exclude these high-churn files until this issue is addressed: https://github.com/JohnnyMorganz/StyLua/issues/829
  ```
  test/functional/ui/decorations_spec.lua  | 3560 ++++++++++++++++++++++++++++++++++++----------------
  test/functional/ui/float_spec.lua        | 5826 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++-------------------------
  test/functional/ui/multigrid_spec.lua    | 1349 ++++++++++++++------
  ```
- Make surgical changes to these files (or add `stylua: ignore` in some
  small scopes) to improve the result:
  ```
  test/functional/vimscript/msgpack_functions_spec.lua | 1414 +++++++++++++++------
  test/functional/api/buffer_spec.lua                  | 1389 +++++++++++----------
  test/functional/api/vim_spec.lua                     | 2740 +++++++++++++++++++++++-----------------
  ```
- These "high churn" files are NOT excluded because the changes are
  largely an improvement:
  ```
  test/functional/plugin/lsp_spec.lua      | 2198 ++++++++++++++++++---------------
  test/functional/plugin/shada_spec.lua    | 4078 +++++++++++++++++++++++++++++++++++-------------------------
  test/functional/ui/cmdline_spec.lua      | 1199 +++++++++++-------
  test/functional/ui/popupmenu_spec.lua    | 1267 +++++++++++--------
  test/functional/ui/messages_spec.lua     | 1643 +++++++++++++++---------
  ```
- TODO: how to check "all directories"? With `GLOB_DIRS *` and `/.deps/` (or  `.deps/`) in `.styluaignore`, Lua code in `.deps/` is still checked...

justinmk commented on CMakeLists.txt:
```diff
@@ -226,16 +226,16 @@ add_glob_target(
   GLOB_DIRS runtime/ scripts/ src/ test/
   GLOB_PAT *.lua
-  TOUCH_STRATEGY SINGLE)
+  TOUCH_STRATEGY PER_DIR)
```
Trying this

dundargoc replied:
Lol, I was fonna suggest we remove `PER_DIR` since we don't use it, and trying to keep `--target help` clean feels like a pipe dream at this point in time.

justinmk replied:
Up to you. But generating thousands of touch-files (and targets) seems excessive so I thought it's worth seeing if we can be less noisy, if the computation is cheap.

justinmk commented on CMakeLists.txt:
```diff
@@ -226,16 +226,16 @@ add_glob_target(
   COMMAND ${STYLUA_PRG}
   FLAGS --color=always --check --respect-ignores
-  GLOB_DIRS runtime/ scripts/ src/ test/unit/
+  GLOB_DIRS runtime/ scripts/ src/ test/
```
I tried `GLOB_DIRS *` , but then `make lintlua` checks `.deps`, even if `/.deps/` (or `.deps/`) is in `.styluaignore`.

dundargoc replied:
I have a suspicion this could be relatednto the implementation in add_glob_target rather than a provlem in stylua. Unsure though.

dundargoc replied:
I take back what I said lol, after experiment around I think this is a stylua problem. It's not a big deal IMO. I'd say maybe slap on a comment and call it a day. Up to you though.

dundargoc commented on CMakeLists.txt:
```diff
@@ -246,7 +246,7 @@ add_glob_target(
   GLOB_DIRS scripts
   GLOB_PAT *.sh
-  TOUCH_STRATEGY SINGLE)
+  TOUCH_STRATEGY PER_DIR)
```
`lintsh` is basically the "ideal" candidate for `SINGLE`, so if we're not using it here it makes me question its efficacy. Maybe we should just remove the `SINGLE` option and just use `PER_DIR` since it roughly does what `SINGLE` does?

dundargoc commented:
I can't even approve because the diff is too big and breaks github app, consider this my approval:

👍

justinmk commented:
> cf any file which makes heavy use of screen tests, like `ui/bufhl_spec.lua` for "simple" cases and `ui/float_spec.lua` for more complex ones.

For now I'll add those and other high-churn files to the exclude-list.

justinmk commented:
See updated description

justinmk commented on test/functional/vimscript/msgpack_functions_spec.lua:
```diff
@@ -20,17 +20,18 @@ describe('msgpack*() functions', function()
   -- 5 identical shada files.
+  -- stylua: ignore
   local big_obj = {
     1, 1436711454, 78, {
```
This is data so added `stylua: ignore`. Nice thing about `stylua: ignore` is that it is local to the scope.

justinmk committed:
build: enable lintlua for test/ dir

Problem:
Not all Lua code is checked by stylua. Automating code-style is an
important mechanism for reducing time spent on accidental
(non-essential) complexity.

Solution:
- Enable stylua for entire `test/` directory.
- Exclude these high-churn files until this issue is resolved: https://github.com/JohnnyMorganz/StyLua/issues/829
  ```
  test/functional/ui/decorations_spec.lua  | 3560 ++++++++++++++++++++++++++++++++++++----------------
  test/functional/ui/float_spec.lua        | 5826 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++-------------------------
  test/functional/ui/multigrid_spec.lua    | 1349 ++++++++++++++------
  ```
- Make surgical changes to these files (or add `stylua: ignore` in some
  small scopes) to improve the result:
  ```
  test/functional/vimscript/msgpack_functions_spec.lua | 1414 +++++++++++++++------
  test/functional/api/buffer_spec.lua                  | 1389 +++++++++++----------
  test/functional/api/vim_spec.lua                     | 2740 +++++++++++++++++++++++-----------------
  ```
- These "high churn" files are NOT excluded because the changes are
  largely an improvement:
  ```
  test/functional/plugin/lsp_spec.lua      | 2198 ++++++++++++++++++---------------
  test/functional/plugin/shada_spec.lua    | 4078 +++++++++++++++++++++++++++++++++++-------------------------
  test/functional/ui/cmdline_spec.lua      | 1199 +++++++++++-------
  test/functional/ui/popupmenu_spec.lua    | 1267 +++++++++++--------
  test/functional/ui/messages_spec.lua     | 1643 +++++++++++++++---------
  ```
- TODO: how to check "all directories"? With `GLOB_DIRS *` and `/.deps/` (or
  `.deps/`) in `.styluaignore`, Lua code in `.deps/` is still checked...

justinmk commented:
This is ready I think. ok @bfredl ?

justinmk committed:
refactor: format test/*

wookayin commented on .styluaignore:
```diff
@@ -1,9 +1,12 @@
-/test/functional
+test/functional/ui/decorations_spec.lua
+test/functional/ui/float_spec.lua
+test/functional/ui/multigrid_spec.lua
```
Alternatively, we could also consider using `-- stylua: ignore start` at the beginning of each file rather than listing files to exclude here in `.styluaignore`, which might make it a bit easier for autoformatters (stylua) to correctly recognize files to ignore.

dundargoc replied:
> which might make it a bit easier for autoformatters (stylua) to correctly recognize files to ignore

Eh, I don't think we need to take this into account. Let's just consider this a minor bug that will probably be fixed soon.

Assuming both methods work equally well: I don't mind either, so if anyone has a strong preference for either one I'm fine to go with that. I guess a nice thing with ignoring with `.styluaignore` is that all the ignored files are neatly in one single place. Another nice thing is that we don't need to modify vendored files such as `/runtime/lua/vim/re.lua`. So I think I'd lean towards keeping the files in `.styluaignore`.

Tbh this is not really a big issue. If stylua happens for format `build` and whatnot isn't *that* big of a problem. It's mostly about reducing the maintenance of cmake files but it's so minor that I don't think we need to put in more time into this IMO.

justinmk replied:
> If stylua happens for format `build` and whatnot isn't _that_ big of a problem

`lintlua` complaining about `.deps/*` can be quite noisy.

dundargoc replied:
Wut? lintlua does not lint `.deps`.
